lib/airbyte/base_client.rb

```diff
@@ -1,96 +1,51 @@
-
-    attr_reader :body, :message, :status_code
-
-    def initialize(msg, code, body)
-      super(msg)
-      @body = body
-      @message = msg
-      @status_code = code
-    end 
-
-  end
-
-  class InputValidationError < StandardError
-
-    attr_reader :body, :status_code, :message
-
-    def initialize(msg, code, body)
-      super(msg)
-      @body = body
-      @message = msg
-      @status_code = code
-    end
-
-  end
-
-  class ObjectNotFoundError < StandardError
-
-    attr_reader :body, :status_code, :message
-
-    def initialize(msg, code, body)
-      super(msg)
-      @body = body
-      @message = msg
-      @status_code = code 
-    end
-
-  end
-
-  class ConnectionError < StandardError
-  end
-
   class BaseClient
-
-    def handle_request(url, http_verb: :post, params: nil, body: nil)
-      begin
-        result = nil
-        Airbyte.connection.with do |faraday_connection|
-          result = faraday_connection.send(http_verb.to_s) do |request|
-            request.url url
-            request.headers['Content-Type'] = 'application/json'
-            request.params = params unless params.nil?
-            request.body = body.to_json unless body.nil?
-            request.options[:timeout] = 600
-          end
-        end
-        handle_result(result)
-      rescue Faraday::TimeoutError, Faraday::ConnectionFailed => ex
-        raise ConnectionError.new("Airbyte connection error: #{ex.message}")
+    def do_handle_request(url, http_verb, params, body)
+      result = nil
+      faraday_connection = establish_connection
+      result = faraday_connection.send(http_verb.to_s) do |request|
+        request.url url
+        request.headers['Content-Type'] = 'application/json'
+        request.params = params unless params.nil?
+        request.body = body.to_json unless body.nil?
+        request.options[:timeout] = 600
       end
+      handle_result(result)
     end
 
-    def handle_request_with_headers(url, http_verb, params: nil, body: nil)
-      raise StandardError.new("Unstubbed Airbyte call from test") if Rails.env.test?
-      result = nil
-      Airbyte.connection.with do |faraday_connection|
-        result = faraday_connection.send(http_verb.to_s) do |request|
-          request.url url
-          request.headers['Content-Type'] = 'application/json'
-          request.params = params unless params.nil?
-          request.body = body.to_json unless body.nil?
-          request.options[:timeout] = 600
+    def handle_request(url, http_verb: :post, params: nil, body: nil, max_retries: 5)
+      retry_count = 0
+      begin
+        do_handle_request(url, http_verb, params, body)
+      rescue Faraday::TimeoutError, Faraday::ConnectionFailed => ex
+        retry_count += 1
+        if retry_count <= max_retries
+          sleep(rand(5))
+          #$statsd.increment("airbyte_client.http.error.retry")
+          retry
         end
+        #$statsd.increment("airbyte_client.http.error.retry_exhausted")
```

Comment:
let's keep statsd here

Reply:
should we do rollbar also when retry exhausts ? Just checking as this should not cause unwanted spams

Reply:
we could add rollbar, somehow statsd was not working in gem, need to check it hence, I kept commented as placed holders.

Reply:
let's add rollbar then. And let's check statsd as well